desc/integrals/_interp_utils.py

```diff
@@ -0,0 +1,754 @@
+
+    """
+    lobatto = bool(lobatto)
+    errorif(lobatto, NotImplementedError, "JAX hasn't implemented type 1 DCT.")
+    a = cheb_from_dct(dct(f, type=2 - lobatto, axis=axis), axis) / (
+        f.shape[axis] - lobatto
+    )
+    return idct_non_uniform(xq, a, f.shape[axis], axis)
+
+
+def idct_non_uniform(xq, a, n, axis=-1):
+    """Evaluate discrete Chebyshev transform coefficients ``a`` at ``xq`` ∈ [-1, 1].
+
+    Parameters
+    ----------
+    xq : jnp.ndarray
+        Real query points where interpolation is desired.
+        Shape of ``xq`` must broadcast with shape ``np.delete(a.shape,axis)``.
+    a : jnp.ndarray
+        Discrete Chebyshev transform coefficients.
+    n : int
+        Spectral resolution of ``a``.
+    axis : int
+        Axis along which to transform.
+
+    Returns
+    -------
+    fq : jnp.ndarray
+        Real function value at query points.
+
+    """
+    # |a| << |xq|, so move a instead
+    a = jnp.moveaxis(a, axis, -1)
+    # Same as Clenshaw recursion ``chebval(xq,a,tensor=False)`` but better on GPU.
+    n = jnp.arange(n)
+    return jnp.linalg.vecdot(jnp.cos(n * jnp.arccos(xq)[..., jnp.newaxis]), a)
+
+
+def _fourier(grid, f, is_reshaped=False):
+    """Transform to DESC spectral domain.
+
+    Parameters
+    ----------
+    grid : Grid
+        Tensor-product grid in (ρ, θ, ζ) with uniformly spaced nodes
+        (θ, ζ) ∈ [0, 2π) × [0, 2π/NFP).
+        Preferably power of 2 for ``grid.num_theta`` and ``grid.num_zeta``.
+    f : jnp.ndarray
+        Function evaluated on ``grid``.
+
+    Returns
+    -------
+    a : jnp.ndarray
+        Shape (..., grid.num_theta // 2 + 1, grid.num_zeta)
+        Complex coefficients of 2D real FFT of ``f``.
+
+    """
+    if not is_reshaped:
+        f = grid.meshgrid_reshape(f, "rtz")
+    # real fft over poloidal since usually M > N
+    return rfft2(f, axes=(-1, -2), norm="forward")
+
+
+# Warning: method must be specified as keyword argument.
+interp1d_vec = jnp.vectorize(
+    interp1d, signature="(m),(n),(n)->(m)", excluded={"method"}
+)
+
+
+@partial(jnp.vectorize, signature="(m),(n),(n),(n)->(m)")
+def interp1d_Hermite_vec(xq, x, f, fx, /):
+    """Vectorized cubic Hermite spline."""
+    return interp1d(xq, x, f, method="cubic", fx=fx)
+
+
+def polyder_vec(c):
+    """Coefficients for the derivatives of the given set of polynomials.
+
+    Parameters
+    ----------
+    c : jnp.ndarray
+        Last axis should store coefficients of a polynomial. For a polynomial given by
+        ∑ᵢⁿ cᵢ xⁱ, where n is ``c.shape[-1]-1``, coefficient cᵢ should be stored at
+        ``c[...,n-i]``.
+
+    Returns
+    -------
+    poly : jnp.ndarray
+        Coefficients of polynomial derivative, ignoring the arbitrary constant. That is,
+        ``poly[...,i]`` stores the coefficient of the monomial xⁿ⁻ⁱ⁻¹,  where n is
+        ``c.shape[-1]-1``.
+
+    """
+    return c[..., :-1] * jnp.arange(c.shape[-1] - 1, 0, -1)
+
+
+def polyval_vec(*, x, c):
```

Comment:
what is the difference btwn this one and the one in `basis.py`? the basis one must have all the same poly degree?

Reply:
that one has loops